### ISSUE TITLE
Derivation path for Copay 1.0.1

### ISSUE THREAD

**Issue by lukem512** [bitpay/copay#3930]
Hello,

I was wondering what the derivation path schema used for Copay 1.0.1 (commit ea818f9bf0dccf6b5152cdd5046a33ff60fc1fe2) is, and how can I find out some more information about this? I have found @maraoz's document entitled [bip-NNNN](https://github.com/maraoz/bips/blob/master/bip-NNNN.mediawiki) that seems to give some clue as the development direction (I don't know if this is still valid?) but it seems incomplete - crucially in that the `NNNN` was never defined.

I am specifically interested in the user's default wallet, which, in the case of the fork I am working with, is a `1-of-1 multisig`. I would like to understand the general schema too, however, so I can derive any private keys that have been used by the user.

Any help will be appreciated.
Thanks

**Comment by dabura667** [comment]
https://bip32jp.github.io/english/copay_decrypt.html

I made this site for the old Copay. Check the javascript to get a hint.

tl;dr

1. Copay wallets are encrypted / decrypted using sjcl (https://bitwiseshiftleft.github.io/sjcl/)
2. The xprv in the wallet is the master key (m in BIP32 terms)
3. The xpubs of all participants (the xpub strings in the wallet) are of m/45'
4. Receive addresses are m/45'/2147483647/0/k and the change addresses are m/45'/2147483647/1/k where k is the key index starting from 0.
5. Copay did not limit addresses (gap limit) so every time a new device would see the open app, they would increase the index forever, and BWS would just keep track of all the used addresses.
6. Remember that the xpubs are of m/45', so when deriving keys from them you must use m/2147483647/0/k (where m represents the xpub string) as most BIP32 sites derive based off of what you input as "m".
7. The pubkey are sorted canonically. so if you have a 2 of 3, and you generate the 3 pubkeys for a given index, you must order them left to right ascending in the redeemscript.

That's about it.

As long as you know how to derive keys from BIP32 extended keys and how to make your own 1 of 1 p2sh multisig redeemscript etc. you should be able to recover funds.

**Comment by dabura667** [comment]
https://coinb.in/   < this will let you generate and send transactions based off of mutlisigs by inputting your own hex pubkeys.

https://bip32jp.github.io/english/  < This will let you derive the keys from the xprv and xpubs. The site is primarily based around BIP39, but if you paste a xpub or xprv into the BIP32 Extended Key box, it will ignore all the data above it (like entropy and BIP39 phrases) and let you generate arbitrary paths from the inputs below it.

With these two tools you should be able to send a transaction, and my decryption tool I posted above should tell you your balance and which addresses have balances.

**Comment by dan-da** [comment]
Also I made a tool for deriving HD wallet addresses that supports both old and current Copay.

CLI:  https://github.com/dan-da/hd-wallet-addrs
Web Form: https://mybitprices.info/hd-wallet-addrs.html

**Comment by lukem512** [comment]
Thanks @dabura667 and @dan-da for your help, particularly on the derivation path. @dan-da, what scheme did you use for finding all the wallets, did you simply scan them until one wasn't seen on the network, or did you pick an arbitrary (large) number and stop there?

**Comment by dan-da** [comment]
@lukem512 I think you mean finding all the addresses, not wallets.    The code derives addresses sequentially in batches and then checks the blockchain to determine if each address has received funds or not.  If it finds 20 in a row that have not received funds, it stops and reports only those that did receive funds.   20 is the default gap limit as defined in one of the bips, but it is a flag and can be set to anything.    Also the program has a flag for generating N addresses, regardless if used or not.  more details in the readme, or the code itself.

**Comment by lukem512** [comment]
Ah yes I do! It's been a long day... Thanks, that seems like a good strategy, I think I'll adopt something similar. I'm just checking your code out now. Thank you!

**Comment by dabura667** [comment]
Copay old versions did not respect gap limits.

Checking only 20 addresses could possibly leave you without discovering funds.

I believe every time you looked at the receive screen it would generate a new address, so if you looked at it a lot you could have a huge gap of like 30 or something.

**Comment by dan-da** [comment]
@dabura667 thx for the info.   I've added a note to the README for hd-wallet-addrs.

I'm curious though, how did old Copay import a wallet from seed words?   This process requires wallet discovery and so it must have stopped when it reached some particular gap limit, no?  If not 20, what?

maybe it just means that old Copay could generate wallets that it could not itself restore fully by that method?

I seem to recall seeing some discussion of these types of issues a while back...

**Comment by dabura667** [comment]
1. Old copay wallets did not use seed words.
2. BWS would store an "addresses" list on the wallet's database entry that has a list of all addresses shown to the user. It also had an index value telling what the highest index value that has been included in the addresses list.

This means the info on BWS server was "technically" necessary for recovery.

This has since been fixed, and current versions respect the gap limit of 20.

**Comment by lukem512** [comment]
Hi all, thanks again for your earlier advice. I have created a tool that will search `m/2147483647/p/k` where `p = {0,1}` and `k = {0,1,2,...} and finds _most_ of the transactions from my wallet, however, there seem to be a few transactions that it cannot find - even when I generate `k`s up to `1000`. Is there ever a situation where the derivation path changes from this - for example, where `p` changes?

**Comment by matiu** [comment]
Yes, 2147483647 is the 'shared' copayer address. If your wallet is *really* old, some address could be derived using a copayer index (0-1-2-N in a M-N wallet).

**Comment by dan-da** [comment]
interesting.   so what determines if 2147483647 is used or copayer index is used?

Is it the case that copayer index was used up until version x.x.x and 2147483647 after, up until 1.6.2 when bip 44 was adopted?     If so:

 1) what version was the last to use copayer index?
 2) Given a wallet with both types of addresses, what algo should be used to discover the addresses?
     I guess it would be a matter of searching up to gap limit (k) for:
        m/2147483647/{0,1}/k
        m/{0,1,2,...}/{0,1}/k

        And stop at the first copayer index that has 0 received tx within gap limit.

        That's potentially a lot of lookups.

**Comment by matiu** [comment]
See https://github.com/bitpay/copay#wallet-export-format

> 1) what version was the last to use copayer index?

Only versions prior to BWS  (ie: peer-to-peer using webRTC) used other that the shared copayer Index. (Beta versions prior to v0.10 https://github.com/bitpay/copay/releases/tag/v0.10.0  ~ Apr 11, 2015).

 > 2) Given a wallet with both types of addresses, what algo should be used to discover the addresses?

you approach is correct. That is the way Copay scan wallets now, JIC.

**Comment by dabura667** [comment]
Wasn't the index supposed to be staggered for each wallet?

Ie. A 2 of 3 wallet would have one person use m/0/x/k and another m/1/x/k and the last one m/2/x/k?

I thought that was how BIP45 was supposed to work.

afaik you have to sort the xpubs by their contained ecdsa pubkey represented in hex and sorted in order lexigraphically in order to determine which one should be 0, 1, 2 etc.

BIP45 should have the details, but if you can't figure it out it should be trivial to try all combinations.

**Comment by lukem512** [comment]
Thanks for the advice, I found the lost `1-of-1` addresses in the `m/45'/0/x/k` path.

**Comment by matiu** [comment]
@dabura667 

> Wasn't the index supposed to be staggered for each wallet?

Yes, that is correct. But when the address is generated at BWS (not independently / offline in a Copayer) we thought it would be better to use the shared index.
